### PR TITLE
Add FormatterCollection<T> with RemoveType<T> method

### PR DESCRIPTION
**Pull request by Krusen**
#2945

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.Formatters
+{
+    public class FormatterCollection<TFormatter> : List<TFormatter>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs]: I'd really prefer that this inherit from `Collection<T>` like in the sample you provided. `List<T>` exposes extra surface area that we have to assume users take a dependency on.

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs]: Also, doc header on the class.

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs]: And if you really still want to be able to call `RemoveAll` you can use this constructor: https://msdn.microsoft.com/en-us/library/ms132401(v=vs.110).aspx and keep a reference to the list you create.

**Comment by Krusen** [comment]
Changed it back to my original proposal using `Collection<T>` and LINQ and added a little documentation on the class.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    public class FormatterCollection<TFormatter> : Collection<TFormatter>
+    {
+
+        /// <summary>
+        /// Removes all formatters of the specified type.
+        /// </summary>
+        /// <typeparam name="T">The type to remove.</typeparam>
+        public void RemoveType<T>() where T : TFormatter
+        {
+            var formatters = this.OfType<T>().ToList();
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/FormatterCollection.cs]:
```
            for (int i = Count - 1; i >= 0; i--)
            {
                var formatter = this[i];
                if(formatter is T)
                {
                    Remove(formatter);
                }
            }
```

**Comment by Krusen** [comment]
Removed LINQ usage as proposed by @davidfowl.

**Comment by davidfowl** [comment]
@Krusen can you add a test?

**Comment by Krusen** [comment]
@davidfowl Yes, of course. I have added a test. Let me know if it needs more tests or anything else.

**Comment by rynowak** [comment]
:shipit: - looks good to me. I'll merge it shortly.

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/FormatterCollectionTests.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        {
+            public override Task WriteResponseBodyAsync([NotNull] OutputFormatterContext context)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        private class AnotherTestOutputFormatter : OutputFormatter
+        {
+            public override Task WriteResponseBodyAsync([NotNull] OutputFormatterContext context)
```
Comment (dougbu) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/FormatterCollectionTests.cs]: @rynowak could you include a commit to remove the `[NotNull]` use in test code? Not something we've done before and now's an odd time to start.

**Comment by dougbu** [comment]
@rynowak also probably better to hold off pushing 'til we get Travis clean again.

File: test/Microsoft.AspNet.Mvc.Core.Test/Formatters/FormatterCollectionTests.cs (hunk 3 above)
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Core.Test/Formatters/FormatterCollectionTests.cs]: Sure

**Comment by rynowak** [comment]
Yeah, I'm looking at travis now. I suspect their `RuntimeHelpers.GetHashCode` doesn't handle null gracefully.

**Comment by rynowak** [comment]
Merged: ec18b3512307f7cac66c26c0668b9bca05b52101 thanks!

**Pull request closed without merging**